packages/server/src/queries/complex/portfolio/allocation.ts

```diff
@@ -0,0 +1,184 @@
+  const unclaimedRewardsCap = +categories["unclaimed-rewards"].capitalization;
+  const pooledCap = +categories["pooled"].capitalization;
+
+  const totalCap =
+    userBalancesCap +
+    stakedCap +
+    unstakingCap +
+    unclaimedRewardsCap +
+    pooledCap;
+
+  return [
+    {
+      key: "available",
+      percentage: (userBalancesCap / totalCap) * 100,
+      amount: userBalancesCap,
+    },
+    {
+      key: "staked",
+      percentage: (stakedCap / totalCap) * 100,
+      amount: stakedCap,
+    },
+    {
+      key: "unstaking",
+      percentage: (unstakingCap / totalCap) * 100,
+      amount: unstakingCap,
+    },
+    {
+      key: "unclaimedRewards",
+      percentage: (unclaimedRewardsCap / totalCap) * 100,
+      amount: unclaimedRewardsCap,
+    },
+    {
+      key: "pooled",
+      percentage: (pooledCap / totalCap) * 100,
+      amount: pooledCap,
+    },
+  ];
+}
+
+async function getAssets(categories: Categories, assetLists: AssetList[]) {
+  const totalAssets = categories["total-assets"];
+  const totalCap = +totalAssets.capitalization;
+
+  // Get top 5 assets by cap value
+  const sortedAccountCoinsResults =
+    totalAssets?.account_coins_result?.sort(
+      (a: AccountCoinsResult, b: AccountCoinsResult) =>
+        +b.cap_value - +a.cap_value
+    ) || [];
+
+  const top5AccountCoinsResults = sortedAccountCoinsResults.slice(0, 5);
+
+  const assets = top5AccountCoinsResults.map((asset: AccountCoinsResult) => {
+    const assetFromAssetLists = getAsset({
+      assetLists,
+      anyDenom: asset.coin.denom,
+    });
+
+    return {
+      key: assetFromAssetLists.coinDenom,
+      percentage: (+asset.cap_value / totalCap) * 100,
+      amount: +asset.cap_value,
+    };
+  });
+
+  const otherAssets = sortedAccountCoinsResults.slice(5);
+  const otherAmount = otherAssets.reduce(
+    (sum: number, asset: AccountCoinsResult) => sum + +asset.cap_value,
+    0
+  );
+  const otherPercentage = (otherAmount / totalCap) * 100;
+
+  const other = {
+    key: "Other",
+    percentage: otherPercentage,
+    amount: otherAmount,
+    color: "bg-osmoverse-500",
+  };
+
+  return [...assets, other];
+}
+
+async function getAvailable(categories: Categories, assetLists: AssetList[]) {
+  const userBalances = categories["user-balances"];
+  const totalCap = +userBalances.capitalization;
+
+  // Get top 5 assets by cap value
+  const sortedAccountCoinsResults =
+    userBalances?.account_coins_result?.sort(
+      (a: AccountCoinsResult, b: AccountCoinsResult) =>
+        +b.cap_value - +a.cap_value
+    ) || [];
+
+  const top5AccountCoinsResults = sortedAccountCoinsResults.slice(0, 5);
+
+  const assets = top5AccountCoinsResults.map((asset: AccountCoinsResult) => {
+    const assetFromAssetLists = getAsset({
+      assetLists,
+      anyDenom: asset.coin.denom,
+    });
+
+    return {
+      key: assetFromAssetLists.coinDenom,
+      percentage: (+asset.cap_value / totalCap) * 100,
+      amount: +asset.cap_value,
```

Comment:
It could simplify all future client side logic if we directly return `percentage` as `RatePretty` and `fiatValue` as `PricePretty`

Later, we may want a server side mechanism to convert prices to other fiat currencies.

Also, the client side implementation is simpler.

In general, from my experience the more complexity we can push to the server, the better.